### DISCUSSION TITLE
Toward a common API

### DISCUSSION THREAD

**Discussion by HajimeKawahara** [HajimeKawahara/exojax#257] [Ideas]
First, I check what quantities from `moldb` are used in ExoJAX retrieval.

## ExoMol
Using [a retrieval w/ CH4](https://github.com/HajimeKawahara/exojax/blob/develop/tests/manualtest/reverse/reverse_methane.py) as an example, I am checking which part should be replaced by a common API.
### init_modit
[here](https://github.com/HajimeKawahara/exojax/blob/6997e0b5d9e11406a695c426299fda74b17b8396/tests/manualtest/reverse/reverse_methane.py#L72)
```python
cnu, indexnu, R, pmarray = initspec.init_modit(mdbCH4.nu_lines, nus)
```
- [x] nu_lines, looks easy to be replaced, already implemented? 

### in setdgm_exomol, exomol
[here](https://github.com/HajimeKawahara/exojax/blob/6997e0b5d9e11406a695c426299fda74b17b8396/tests/manualtest/reverse/reverse_methane.py#L83) and [here](https://github.com/HajimeKawahara/exojax/blob/6997e0b5d9e11406a695c426299fda74b17b8396/tests/manualtest/reverse/reverse_methane.py#L100)
```python
dgm_ngammaL = setdgm_exomol(mdbCH4, fT, Parr, R, molmassCH4, res, T0_test, alpha_test)
```
or
```python
SijM_CH4, ngammaLM_CH4, nsigmaDl_CH4 = modit.exomol(
        mdbCH4, Tarr, Parr, R, molmassCH4)
```

[The destination is modit.exomol](https://github.com/HajimeKawahara/exojax/blob/6997e0b5d9e11406a695c426299fda74b17b8396/src/exojax/spec/modit.py#L256)
```
    qt = vmap(mdb.qr_interp)(Tarr)
    SijM = jit(vmap(SijT, (0, None, None, None, 0)))(
        Tarr, mdb.logsij0, mdb.dev_nu_lines, mdb.elower, qt)
    gammaLMP = jit(vmap(gamma_exomol, (0, 0, None, None)))(
        Parr, Tarr, mdb.n_Texp, mdb.alpha_ref)
    gammaLMN = gamma_natural(mdb.A)
    gammaLM = gammaLMP+gammaLMN[None, :]
    ngammaLM = gammaLM/(mdb.dev_nu_lines/R)
```
- [ ] (relative) partition function, `qr_interp`, need to be considered
- [ ] logsij0, looks relatively easy to be replaced
- [x] nu_lines in jnp.array, already almost implemented 
- [x] elower,  already implemented 
- [x] pressure broadening parameters, `n_Texp, alpha_ref`, already implemented?
- [x] Einstein coefficient, already implemented 

### CIA
[here](https://github.com/HajimeKawahara/exojax/blob/6997e0b5d9e11406a695c426299fda74b17b8396/tests/manualtest/reverse/reverse_methane.py#L107)
```python
    dtaucH2H2 = dtauCIA(nus, Tarr, Parr, dParr, vmrH2, vmrH2,
                        mmw, g, cdbH2H2.nucia, cdbH2H2.tcia, cdbH2H2.logac)
```
- [ ] CIA/HITRAN
I think we should postpone CIA meanwhile. But it should be included in the common API in the future?

## HITEMP
I use [this example](https://github.com/HajimeKawahara/exojax/blob/develop/tests/manualtest/reverse/reverse_CO_HITEMP.py)
Essentially same as the Exomol case but,

### modit.hitran
[here](https://github.com/HajimeKawahara/exojax/blob/6997e0b5d9e11406a695c426299fda74b17b8396/tests/manualtest/reverse/reverse_CO_HITEMP.py#L102)

```python
SijM_CO, ngammaLM_CO, nsigmaDl_CO = modit.hitran(
        mdbCO, Tarr, Parr, Parr*VMR_CO, R, molmassCO)
```

The destination is [modit.hitran].(https://github.com/HajimeKawahara/exojax/blob/6997e0b5d9e11406a695c426299fda74b17b8396/src/exojax/spec/modit.py#L300)

```python
    qt = vmap(mdb.Qr_line_HAPI_jax)(Tarr)
    SijM = jit(vmap(SijT, (0, None, None, None, 0)))(
        Tarr, mdb.logsij0, mdb.dev_nu_lines, mdb.elower, qt)
    gammaLMP = jit(vmap(gamma_hitran, (0, 0, 0, None, None, None)))(
        Parr, Tarr, Pself, mdb.n_air, mdb.gamma_air, mdb.gamma_self)
    gammaLMN = gamma_natural(mdb.A)
    gammaLM = gammaLMP+gammaLMN[None, :]
    ngammaLM = gammaLM/(mdb.dev_nu_lines/R)
    nsigmaDl = normalized_doppler_sigma(Tarr, molmass, R)[:, jnp.newaxis]
```

- [ ] (relative) partition function, `Qr_line_HAPI_jax`, need to be considered
- [ ] logsij0, looks relatively easy to be replaced
- [x] nu_lines in jnp.array, already almost implemented 
- [x] elower,  already implemented 
- [x] pressure broadening parameters, `n_Texp, alpha_ref`, already implemented?
- [x] Einstein coefficient, already implemented 

## short summary
It looks easy to be replaced except for the partition function stuff and CIA.

**Comment by erwanp** [comment]
In Radis's https://github.com/radis/radis/blob/e317c994451d3afa02b1fa63d90ae0eb546f36ed/radis/io/exomol.py#L925  I added an Example of an Exojax-like syntax, creating a ``mdbExoMol` class holding jax arrays as Attributes (based on DataFrame loading), so it can be used as a drop-in replacement 

The example also shows how to : 
1. Compute an Exomol CH4 spectrum with Radis
2. Load Exomol CH4 Jax arrays ready for an Exojax computation, by downloading the files to the .database local folder 
3. Load Exomol CH4 Jax arrays ready for an Exojax computation, by loading the files from the Radis's database of (1)

For 2. and 3., a `local_databases=` parameter allows to easily switch in between any local folder (to save on download & caching times if comparing the two codes) 

(note: for an actual drop-in replacement; the `mdbExoMol` class still need to be given Partition function calculation methods, etc.  ) 

---

Can be tried in https://github.com/radis/radis/pull/465 ; running radis/io/exomol.py

**Reply by HajimeKawahara** [reply]
Cool! I'll take your code as a reference.

**Comment by HajimeKawahara** [comment]
Just a thought on output form in `fetch_exomol`

### GPU memory allocation (jnp.array)
In ExoJAX, we need to care more about GPU memory than DRAM.

In the current opacity calculators,  LPF and MODIT, we use
- mdb.logsij0, mdb.dev_nu_lines, mdb.elower, mdb.n_Texp, mdb.alpha_ref (for ExoMol)
- mdb.logsij0, mdb.dev_nu_lines, mdb.elower, mdb.n_air, mdb.gamma_air, mdb.gamma_self (for HITEMP)
in jnp.array, i.e. on GPU memory.

1. we should not put other quantities on GPU memory, so not as jnp.array().

These quantities are proportional to the number of the lines. In ExoJAX, we are now moving in the direction of a wider wavelength range. This will enable us to apply ExoJAX both to wide range high-res spectroscopy such as IRD/Subaru, CRIRES+/VLT and to  low and med resolution spectroscopy (w/ high res if you want), such as HST, JWST, Ariel.  For this purpose, we are now trying two different opacity calculators extending MODIT; one is pseudo line generator (PLG) [(see the PR)](https://github.com/HajimeKawahara/exojax/pull/255) and another is [PreMODIT](https://github.com/HajimeKawahara/exojax/tree/premodit), which pre-compute the LSD then apply the temperature correction. For these methods, we want to put compressed information in GPU memory but others info should not be in GPU memory. 

2. in this case, probably we compress the information using the vaex or pandas form of line info then put compressed info into jnp.array. 

So, probably the output form at `fetch_exomol` should be more flexible.

**Reply by erwanp** [reply]
Using `mdBExomol` or `fetch_exomol` will mostly be preference if you want to retain a class-based database object (lines + partition functions); or function based database object (returning a dictionary + you have to handle partition functions separately). 
In both cases, underlying is a DataFrame structure which yields the lines with [mdb.load()]( https://github.com/radis/radis/blob/e317c994451d3afa02b1fa63d90ae0eb546f36ed/radis/io/exomol.py#L1032) . 
From what I understand, to minimize GPU memory the `columns=` argument of mdb.load() should be `columns=["logsij0", "nu_lines", "elower", "n_Texp", "alpha_ref"]` for ExoMol  ; and equivalent for HITEMP. 

---

(side note on memory management, RADIS 0.11 (October 2021) introduced a Sparse version of the DIT algorithm (https://github.com/radis/radis/pull/386) ; which allowed us to compute extremely large spectra (i.e. full range CO2 HITEMP 5M lines 500 - 10,000 cm-1 on a Radis-Lab instance of only 8 GB RAM. Beforehand we were memory-limited to ranges <1000 cm-1 for approximately the same conditions).  It could help ExoJax scale to wide wavelength ranges. (the idea being that each point of the DIT grid is a full spectrum; but most of these spectra are sparse because the lines with similar lineshape parameters often end up having similar transition-wavenumber ; and many points of the DIT grid are totally empty. See the figures there https://github.com/radis/radis/issues/222 to immediatly understand  ) .  Not sure how it would scale with GPU ; I could try an implementation but unfortunately not before the summer. I can easily guide someone who would try the approach, however! )

**Reply by HajimeKawahara** [reply]
Thanks. For the GPU memory issue, I think there are two sides: One is the case that the lineshape density (LSD) is sparse, i.e. Nline (# of lines) << Ngrid (# of LSD), as you mentioned. Another is, the target of Pre-MODIT, that Nline >> Ngrid, such as methane. Year, in future, we need to care about both cases.

**Reply by HajimeKawahara** [reply]
@erwanp I started to understand the structure of the [PR #465](https://github.com/radis/radis/pull/465). I found `DatabaseManager` is great.
You made a prototype of MdbExomol in ExoJAX that uses radis.io, should be imported in ExoJAX. Is this correct?
I am wondering how we should separate functionalities of a common API from ExoJAX and probably radis specific functionalities. 
Which part in the current radis should be regarded as a common API? 
For me, it looks useful to use radis.io in ExoJAX, but probably, a scope of a common API is wider than that. Technically, it's likely to be difficult to develop without breaking the connection to the radis, if we edit radis.io. What's your thought?

**Reply by HajimeKawahara** [reply]
Some thoughts on a common API:

1. What the common API is in charge of

For instance, `MdbExomol` in [PR 465](https://github.com/radis/radis/pull/465.) should be in ExoJAX because there contains ExoJAX specific functionality.

Potential roles of a common API

- Easy access to line information, sij0, A, logsij0, Elower, .... in  pandas, jax, vaex
- Easy access to the partition functions, in the forms of numpy based and jax based
- Easy access to gamma as a function of T and P, in the forms of numpy and jax based. Note: there are several choices of gamma for metal lines

2. Where to host a common API.

Create another repository for a common API v.s. A module in `radis` as a common API 
Probably, the latter is easier to start because `radis` already has a cool DataFrame structure. But, in this case, I think, we should clarify which module should be regarded as a common API. If we regard `radis.io` as a common API, `radis.io` is going to have two roles; one is the common API of molecular/atomic databases, the other is `io` for radis. I think this makes new development difficult, for instance, importing Krucz/VLAD3  capability from the current ExoJAX implementation to the common API. I am not sure how we can solve this situation. For instance, creating a new module `radis.(common_API_name)` then `radis.io` and `exojax.spec.moldb` call it?

**Reply by erwanp** [reply]
Good questions. 

`radis.io.fetch_hitran, radis.io.fetch_exomol, radis.io.fetch_hitemp` etc. should be radis-specific API; as they contain some RADIS-specific features like Registering the database in the radis.json configuration file; or specific isotope abundance [fetch_exomol uses Earth abundance for consistency with hitran/hitemp whereas default Exomol database does not assume any terrestrial abundance ] 

All of these functions currently build on a subclass of the DatabaseManager. This and its subclass should be our common API. 

ExoJax should build its own functino (equivalent to radis.io.fetch_hitran) or build another subclass (if you like having a class-like-access t o the database like current MdBExomol . 

So we would have (assumes Exojax keeps a class-like approach to the Mdb)
```

COMMON API 
class DatabaseManager             class HitranManager(DatabaseManager)        class ExomolManager(DatabaseManager)
                                                     
RADIS API                         def fetch_hitran()                          def fetch_exomol()                       
                                     mdb = HitranManager()                       mdb = ExomolManager
                                     return mdb.load()                           return mdb.load()

EXOJAX API                        class MdbHitran(HitranManager)              class MdbExomol(Exomol)
                                  def __init__()                              def __init__()
                                     super(MdbExomol).__init__()                 super(MdbExomol).__init__()
                                     # exojax specific init & methods            # exojax specific init & methods

```

What do you think ? 


---
2. You're right we should make things very separate if stored within RADIS.   One way will be to add tests; so ExoJax requirements are tested within RADIS and never break.  Anothr way is to use different folders. Moving things to a `radis.api` folder is a good idea, actually. Plus, when the package grows in maturity in may become independant.

**Reply by HajimeKawahara** [reply]
Thanks! The structure looks great. 
Also, I agree that we start from `radis.api`.

**Reply by erwanp** [reply]
Something related to an API : SPEARS [doi.org/10.1016/j.jqsrt.2021.107958](https://www.sciencedirect.com/science/article/abs/pii/S0022407321004507) is an API released earlier this year ; they provide an universal approach to various databases; and broadening coefficients ; unfortunately their code is not open-source so no one can really use it.

**Reply by erwanp** [reply]
@HajimeKawahara RADIS refactor was done in a new RADIS `add/common-api` branch https://github.com/radis/radis/pull/478  
Everything in `radis.api` should be under ownership of both Radis and Exojax.

**Reply by HajimeKawahara** [reply]
Thanks a lot! I'll make a counterpart interface in ExoMol.

**Comment by HajimeKawahara** [comment]
Started to think about breakdown what to do.
First, 'premodit' corrected moldb.py to separate DRAM and GPU memory (jnp).
See [diff](https://github.com/HajimeKawahara/exojax/pull/265/files#diff-d1df717e61fa3f4e7b44d451d3b200f4e9e1bc0f86d9a40a0163918e6b0d02d9).

**Reply by HajimeKawahara** [reply]
![IMG_20220713_102529](https://user-images.githubusercontent.com/15956904/178629591-089e2437-870e-4e2d-a7b0-e0d9acaf933b.jpg)
